provider/etcd/etcd_client_test.go

```diff
@@ -0,0 +1,272 @@
+	"github.com/prometheus/alertmanager/types"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+)
+
+var (
+	verbose = true
+
+	fakeAlertCounter = 0
+
+	etcdPrefix      = "am/test/alerts-"
+	alertGcInterval = 200 * time.Millisecond
+
+	etcdLogger log.Logger
+)
+
+func init() {
+	if verbose {
+		w := log.NewSyncWriter(os.Stderr)
+		etcdLogger = log.NewJSONLogger(w)
+	} else {
+		etcdLogger = log.NewNopLogger()
+	}
+}
+
+func TestEtcdWriteReadDeleteAlert(t *testing.T) {
+	clus := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+	etcdEndpoints := []string{clus.Members[0].GRPCAddr()}
+
+	marker := types.NewMarker(prometheus.NewRegistry())
+	alerts, err := NewAlerts(context.Background(), marker, alertGcInterval, etcdLogger,
+		etcdEndpoints, etcdPrefix)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// write and read back
+	a1 := fakeAlert()
+	if err := alerts.EtcdClient.Put(a1); err != nil {
+		t.Errorf("etcdPut failed: %s", err)
+	}
+	a2, err := alerts.EtcdClient.Get(a1.Fingerprint())
+	if err != nil {
+		t.Errorf("etcdGet failed: %s", err)
+	}
+	if !alertsEqual(a1, a2) {
+		t.Errorf("Unexpected alert: %s", pretty.Compare(a1, a2))
+	}
+
+	// delete and read back
+	err = alerts.EtcdClient.Del(a1.Fingerprint())
+	if err != nil {
+		t.Errorf("etcdDel failed: %s", err)
+	}
+	_, err = alerts.EtcdClient.Get(a1.Fingerprint())
+	if err == nil {
+		t.Errorf("etcdGet SHOULD HAVE failed")
+	}
+}
+
+func TestEtcdMarshalUnmarshalAlert(t *testing.T) {
+	var str1, str2 string
+	var err error
+	var a1, a2 *types.Alert
+
+	a1 = fakeAlert()
+	if str1, err = MarshalAlert(a1); err != nil {
+		t.Errorf("marshal alert failed: %s", err)
+	}
+	if a2, err = UnmarshalAlert(str1); err != nil {
+		t.Errorf("unmarshal alert failed: %s", err)
+	}
+	if str2, err = MarshalAlert(a2); err != nil {
+		t.Errorf("re-marshal alert failed: %s", err)
+	}
+	if str1 != str2 {
+		t.Error("alert string comparison failed")
+	}
+	if !alertsEqual(a1, a2) {
+		t.Errorf("Unexpected alert: %s", pretty.Compare(a1, a2))
+	}
+}
+
+func TestEtcdRunWatch(t *testing.T) {
+	clus := integration.NewClusterV3(t, &integration.ClusterConfig{Size: 1})
+	defer clus.Terminate(t)
+	etcdEndpoints := []string{clus.Members[0].GRPCAddr()}
+
+	marker := types.NewMarker(prometheus.NewRegistry())
+	alerts, err := NewAlerts(context.Background(), marker, alertGcInterval, etcdLogger,
+		etcdEndpoints, etcdPrefix)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	alerts.EtcdClient.RunWatch(context.Background())
+	iterator := alerts.Subscribe()
+	time.Sleep(100 * time.Millisecond) // wait for subscribe
+
+	// send all of the alerts
+	alertsToSend := []*types.Alert{fakeAlert(), fakeAlert(), fakeAlert()}
+	for _, a := range alertsToSend {
+		if err := alerts.EtcdClient.Put(a); err != nil {
+			t.Errorf("etcdPut failed: %s", err)
+		}
+	}
+
+	// read the alerts back in order
+	index := 0
+	for alert := range iterator.Next() {
```

Comment:
Q: If there are less alerts sent to etcd for some reason, this `Next` will wait forever right?

Reply:
YES.  It will block.